chorus_lib/src/transport/http.rs

```diff
@@ -3,53 +3,71 @@
 use std::thread;
 use std::{collections::HashMap, sync::Arc};
 
+use std::marker::PhantomData;
+
 use retry::{
     delay::{jitter, Fixed},
     retry,
 };
 use tiny_http::Server;
 use ureq::{Agent, AgentBuilder};
 
+use crate::transport::TransportConfig;
+
 use crate::{
-    core::{Portable, Transport},
+    core::{ChoreographyLocation, HList, Member, Portable, Transport},
     utils::queue::BlockingQueue,
 };
 
+type QueueMap = HashMap<String, BlockingQueue<String>>;
+/// A type alias for `TransportConfig`s used for building `HttpTransport`
+pub type HttpTransportConfig<L, Target> = TransportConfig<L, (String, u16), Target, (String, u16)>;
+
 /// The header name for the source location.
 const HEADER_SRC: &str = "X-CHORUS-SOURCE";
 
+/// A wrapper for HashMap<String, (String, u16)>
+#[derive(Clone)]
+pub struct HttpConfig<L: HList> {
+    /// The information about locations
+    pub info: HashMap<String, (String, u16)>,
+    /// The struct is parametrized by the location set (`L`).
+    pub location_set: PhantomData<L>,
+}
+
 /// The HTTP transport.
-pub struct HttpTransport {
+pub struct HttpTransport<L: HList, TLocation> {
     config: HashMap<String, (String, u16)>,
     agent: Agent,
-    queue_map: Arc<HashMap<String, BlockingQueue<String>>>,
     server: Arc<Server>,
     join_handle: Option<thread::JoinHandle<()>>,
+    location_set: PhantomData<L>,
+    queue_map: Arc<QueueMap>,
+    target_location: PhantomData<TLocation>,
 }
 
-impl HttpTransport {
-    /// Creates a new `HttpTransport` instance from the projection target and a configuration.
-    pub fn new(at: &'static str, config: &HashMap<&str, (&str, u16)>) -> Self {
-        let config = HashMap::from_iter(
-            config
-                .iter()
-                .map(|(k, (hostname, port))| (k.to_string(), (hostname.to_string(), *port))),
-        );
-        let locs = Vec::from_iter(config.keys().map(|s| s.clone()));
-
-        let queue_map = {
+impl<L: HList, TLocation: ChoreographyLocation> HttpTransport<L, TLocation> {
+    /// Creates a new `HttpTransport` instance from the configuration.
+    pub fn new<Index>(http_config: &HttpTransportConfig<L, TLocation>) -> Self
```

Comment:
Now that the config is parametrized over the target location, can we just take the ownership of `HttpTransportConfig` instead of taking a reference?